### PR TITLE
start to add an inlining pass

### PR DESCRIPTION
**Pull request by schweitzpgi**
This adds the necessary procedures to perform inlining and a small regression test.

TODO: implement the policy.

### CODE REVIEW

File: flang/include/flang/Optimizer/Transforms/Passes.h [hunk 1]
```diff
@@ -29,14 +32,21 @@ std::unique_ptr<mlir::Pass> createPromoteToAffinePass();
...
 
 /// A pass to convert the FIR dialect from "Mem-SSA" form to "Reg-SSA"
 /// form. This pass is a port of LLVM's mem2reg pass, but modified for the FIR
 /// dialect as well as the restructuring of MLIR's representation to present PHI
 /// nodes as block arguments.
 std::unique_ptr<mlir::Pass> createMemToRegPass();
 
+/// This pass will preform inlining on FIR. If inlining is disabled this returns
+/// a nullptr.
+std::unique_ptr<mlir::Pass> createInlinerPass();
```
Comment (jeanPerier) [inline flang/include/flang/Optimizer/Transforms/Passes.h:45]: The comment on the implementation of this functions describes a different behavior:
> If inlining is disabled, the compiler will abort if an inlining pass is instantiated.

**Comment by rjnw** [comment]
Will adding `CallOpInterface` to `fir.call` do automatic inlining in absence of effects or does the `MemoryEffects` in call op stop that by default?

File: flang/include/flang/Optimizer/Transforms/Passes.h (hunk 1 above)
Comment (schweitzpgi) [reply flang/include/flang/Optimizer/Transforms/Passes.h:45]: Thanks. I'll fix this one.

**Comment by schweitzpgi** [comment]
> Will adding `CallOpInterface` to `fir.call` do automatic inlining in absence of effects or does the `MemoryEffects` in call op stop that by default?

I changed the memory effects annotations to be more like the standard dialect in a previous PR.

Theoretically speaking, effects should thread through calls in general. However, the MLIR model has been evolving in the last couple of months. So it's a bit of an open question wrt FIR.

**Pull request merged**